### ISSUE TITLE
x87 / SSE floating point mismatch for surface extents

### ISSUE THREAD

**Issue by Paril** [yquake2/yquake2#886]
Certain surfaces (mostly ones that aren't aligned on integer boundaries) will round to different values in yquake2 as compared to vanilla engines & light compiler tools. I can't really show an example in-engine since gl_lightmap/r_lightmap isn't implemented, but I can confirm that the issue is still present in this engine (walking over surfaces that have corrupted lightmaps will have stripes of light instead of a flat lightmap surface).

The fix the Q1 community has applied was to cast the texture coordinate calculations to long double, which calculates them as 128-bit double and ends up fixing any found rounding errors from the 80-bit float version.

The lines that would need patching are here:
https://github.com/yquake2/yquake2/blob/9a11251740cccd869614b01bf6e9215db377464b/src/client/refresh/gl1/gl1_model.c#L533
https://github.com/yquake2/yquake2/blob/9a11251740cccd869614b01bf6e9215db377464b/src/client/refresh/gl3/gl3_model.c#L418
https://github.com/yquake2/yquake2/blob/a80f5b6cd7f911370840e5c2dc6f2bf83d0b92f8/src/client/refresh/soft/sw_model.c#L575
https://github.com/yquake2/ref_vk/blob/674e634679f503cdc62ba7cb39ec1002e329eb71/src/vk/vk_model.c#L466

You can see screenshots and a more detailed description of the issue and the work the Quake community had done prior to fix the same bug (Q1 suffers from the same mismatch, and engines have been using the long double fix for like 5+ years) here: https://github.com/skullernet/q2pro/issues/261#issuecomment-1229587645

**Comment by DanielGibson** [comment]
> I can't really show an example in-engine since gl_lightmap/r_lightmap isn't implemented

The gl1 renderer still supports it

**Comment by maraakate** [comment]
Floating point precision and Quake engine bugs strike again.  Interesting fix.  May be a stupid question, but does the precision make a difference in timedemos.  Specifically on limited hardware?  Could be a worthwhile addition to my id engine projects that target DOS and Win9x-era machines, but if it's a performance hit won't be worth it.

**Comment by Paril** [comment]
> > I can't really show an example in-engine since gl_lightmap/r_lightmap isn't implemented
> 
> The gl1 renderer still supports it

Thanks, I thought I had tried that but maybe I didn't, heh.

So, skuller @ q2pro pointed out that the original Q2 vanilla client actually used 24-bit float calculations. The thing is though that the original radiosity tool didn't use this mode, so it would have used the higher precision 80-bit floats - the BSPs are basically baked assuming these 80-bit floats. This means even Q2 vanilla's client would display corrupted lightmaps sometimes (I confirmed that a specific face in rammo1 is broken the same way in vanilla Q2:)

![](https://camo.githubusercontent.com/c3ffb483fbc4d6573e78e79d9a482c6b47f712364a712756ff9417a37503c180/68747470733a2f2f692e696d6775722e636f6d2f73496753564d6f2e706e67)
![](https://camo.githubusercontent.com/f14ba01bb9137d2005f1fa09da1246f381dea96c15c6b1a0f195f9a0a01a05fb/68747470733a2f2f692e696d6775722e636f6d2f69546d4d4461492e706e67)

The Q2 vanilla client displays it the same as the top one, which is 32-bit float calculation. Upping the calculation to long double (and maybe double, but I didn't test it) has it match the qrad3 values, on the bottom.

The result should not affect timedemo, no. This is only done at load time.

**Comment by DanielGibson** [comment]
> Upping the calculation to long double (and maybe double, but I didn't test it)

TBH, if normal double is not enough (when handling values that are normally handled as floats throughout the code), then there most likely is something fundamentally wrong with the calculation

**Comment by Paril** [comment]
They were 80-bit intermediates in qrad3, so 64-bit might not be enough depending on how small the value is - but you're right, double should probably be enough. The calculation is very dumb and it sucks that they relied on it rather than writing the extents directly to the BSP.

**Comment by DanielGibson** [comment]
I generally would like to avoid long double, because it's not available on all platforms (or is the same as normal 64bit double then). I think that applies to ARM, and according to https://docs.microsoft.com/en-us/previous-versions/9cx8xs15(v=vs.140) even to x86 with Visual C++ (?!)

Do you have examples of what specific values in `v->position[i]` and `tex->vecs[j][i]` (and maybe `mins[j]` and `maxs[j]` if that matters) we're dealing with in the broken cases?
Trying to understand why apparently so much precision is needed, when the vertex positions (world coordinates) AFAIK should usually be < +/- 4096

**Comment by Paril** [comment]
Yeah double will work, you can avoid long double. It should be fine.

I will see if I can get this map/bsp over, since it has a ton of mismatches, but in the meantime if you check `rammo1` from Rogue, it has one face (screenshot in the other issue) that has a mismatch. If you calculate both at load time you can print out the mismatch and what the values are that end up coming out of it. It's almost always off by one (then shifted). If you have the LMCTF pack, you can find some in there too; lmctf41 and lmctf43 both have quite a few surfaces that have mismatches.

**Comment by DanielGibson** [comment]
Can you run the `viewpos` console command at the affected place in rammo1 so we can more easily find it?
(BTW, the counter-part of `viewpos` is `teleport`)

**Comment by Paril** [comment]
`878 -77 1895`, yaw of 0, you'll be looking right at it.

Values in console for this one:
![](https://i.imgur.com/tOQz4Q2.png)

**Comment by DanielGibson** [comment]
Thanks!
To make things (even) easier, do you have the diff for those debug prints in your screenshots?

**Comment by Paril** [comment]
This is in Q2PRO, which has it all in one function so it's pretty easy to print out - I don't have a yquake2 adjustment offhand. Basically I just calculated a double version alongside the float one (for the texture coordinates and the mins/maxs), calculated a second extents using the double precision one, and checked it they were equivalent.

**Comment by DanielGibson** [comment]
Thanks!
